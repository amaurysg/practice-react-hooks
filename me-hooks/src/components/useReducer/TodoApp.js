import React, { useEffect, useReducer } from "react";
import TodoList from "./TodoList";
import TodoAdd from "./TodoAdd";
import todoReducer from "./todoReducer";
import "./todoStyle.css";

//Init mode
//Init return is initialState
const init = () => {
  //Here captures values from localStorage
  //captures items from localStore from todos, else capture one array []
  return JSON.parse(localStorage.getItem("todos")) || [];

  //   return [{
  //   id: new Date().getTime(),
  //   desc : 'Either string',
  //   done: false,
  // }]
};

//Initial component
const TodoApp = () => {
  //Here called useReducer
  //receive function reducer , initialState, and init (see init above )
  const [todos, dispatch] = useReducer(todoReducer, [], init);

  //testing state (todos)
  console.log("To Do's", todos);

  ///**** ADD LOCAL STORAGE ****///
  //useEffect for localstorage, when todos change
  //Only add when (todos) change
  useEffect(() => {
    //Add to localStorage
    //Here add to localStorage, we used JSON.stri... because localstorage only read strings
    localStorage.setItem("todos", JSON.stringify(todos));
  }, [todos]);

  ///**** handleDelete ****///
  const handleDelete = (todoId) => {
    console.log(todoId);
    //create action
    //connect dispatch
    dispatch({
      type: "delete",
      payload: todoId,
    });
  };
  ///**** handleToogle ****///
  const handleToogle = (todoId) => {
    dispatch({
      type: "toogle",
      payload: todoId,
    });
  };

  const handleAddToDo = (newToDo) => {
    //Action add, type, payload
    dispatch({
      type: "add",
      payload: newToDo,
    });
  };

  return (
    <div>
      <h1>App ( {todos.length} )</h1>
      <hr />
      <div className="row">
        <div className="col-7">
          <TodoList
            todos={todos}
            handleDelete={handleDelete}
            handleToogle={handleToogle}
          />
        </div>
        <div className="col-5">
          <TodoAdd handleAddToDo={handleAddToDo} />
        </div>
      </div>
    </div>
  );
};

export default TodoApp;
