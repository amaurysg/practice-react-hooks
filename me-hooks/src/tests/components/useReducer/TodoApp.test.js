import { mount, shallow } from "enzyme";
import { act } from "@testing-library/react";
import TodoApp from "../../../components/useReducer/TodoApp";
import demoTodos from "../../fixtures/demoTodos";

describe("Test en component TodoApp", () => {
  // const handleDelete = jest.fn();
  const wrapper = shallow(<TodoApp />);

  //mock localStorage
  //Storage viene con jest
  Storage.prototype.setItem = jest.fn(() => {});

  test("Debe de mostrarte correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Debe agregar un ToDo", () => {
    //mount funciona como shallow, pero tiene mas alcance, m√°s scope
    //cuando necesitamos mas contexto de otros componentes
    const wrapper = mount(<TodoApp />);
    act(() => {
      wrapper.find("TodoAdd").prop("handleAddToDo")(demoTodos[0]);
      wrapper.find("TodoAdd").prop("handleAddToDo")(demoTodos[1]);
    });

    expect(wrapper.find("h1").text().trim()).toBe("App ( 2 )");
    //esperar que se haya llamado dos veces
    expect(localStorage.setItem).toHaveBeenCalledTimes(2);
  });

  test("Que se borre un todo", () => {
    wrapper.find("TodoAdd").prop("handleAddToDo")(demoTodos[0]);
    wrapper.find("TodoList").prop("handleDelete")(demoTodos[0].id);
    expect(wrapper.find("h1").text().trim()).toBe("App ( 0 )");
  });
});
