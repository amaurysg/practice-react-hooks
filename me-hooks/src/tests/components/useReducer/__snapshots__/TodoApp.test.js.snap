// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test en component TodoApp Debe de mostrarte correctamente 1`] = `
<div>
  <h1>
    App ( 
    0
     )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleDelete={[Function]}
        handleToogle={[Function]}
        handleUpdate={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        handleAddToDo={[Function]}
      />
    </div>
  </div>
</div>
`;
